### PR TITLE
[sival] Add test for the entropy source's firmware extract and insert mode

### PR DESCRIPTION
**Pull request by jwnrt**
## Summary

This PR extends the entropy source firmware override observe fifo test to enable "extract and insert" mode.

1. Adds an interrupt handler for the `es_observe_fifo_ready` interrupt which reads 32 words (the FIFO threshold value) from the observe FIFO and writes it back again.
2. Allows configuring the entropy source with extract and insert mode, enabling that interrupt handler.
3. For each of the "single bit" configurations, checks the entropy source is working by generating a set of random numbers and running AES, KMAC, and HMAC.

### CODE REVIEW

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c [hunk 1]
```diff
@@ -103,9 +108,63 @@ static status_t let_observe_fifo_overflow(uint32_t timeout_usec) {
...
+  CHECK(!entropy_src_fifo_has_overflowed(),
+        "observe FIFO overflowed during collection");
+
+  uint32_t buf[32];
+  size_t len = ARRAYSIZE(buf);
+  CHECK_DIF_OK(
+      dif_entropy_src_observe_fifo_nonblocking_read(&entropy_src, buf, &len));
+
+  if (len > 0) {
+    CHECK_DIF_OK(
```
Comment (pamaury) [inline sw/device/tests/entropy_src_fw_observe_many_contiguous.c:131]: Maybe add a comment explaining why it is ok to do this operation in the IRQ handler?

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c [hunk 2]
```diff
@@ -103,9 +108,63 @@ static status_t let_observe_fifo_overflow(uint32_t timeout_usec) {
...
+        dif_entropy_src_observe_fifo_write(&entropy_src, buf, len, &len));
```
Comment (pamaury) [inline sw/device/tests/entropy_src_fw_observe_many_contiguous.c:132]: what happens if the function does not manage to write the whole buffer? In this case, the test will not have pushed contiguous words to the FIFO. I think the code should at least check that possibility and explicitely handle it (error out?).

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c (hunk 1 above)
Comment (jwnrt) [reply sw/device/tests/entropy_src_fw_observe_many_contiguous.c:131]: Do you mean justify why writing to the FIFO will not block the IRQ for too long?

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c (hunk 2 above)
Comment (jwnrt) [reply sw/device/tests/entropy_src_fw_observe_many_contiguous.c:132]: Yeah okay, I've added an error to this test but I'm not sure how a real implementation would deal with that. What do you do if the noise source is still generating bits but there's no room in the FIFO for them? You've got to drop the bits somewhere...

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c [hunk 3]
```diff
@@ -103,9 +108,68 @@ static status_t let_observe_fifo_overflow(uint32_t timeout_usec) {
   return OK_STATUS();
 }
 
+void ottf_external_isr(uint32_t *exc_info) {
```
Comment (h-filali) [inline sw/device/tests/entropy_src_fw_observe_many_contiguous.c:111]: Great idea doing this with an ISR btw!

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c [hunk 4]
```diff
@@ -139,7 +203,8 @@ static status_t entropy_config(
   TRY(dif_entropy_src_configure(
       &entropy_src,
       (dif_entropy_src_config_t){
-          .fips_enable = true,
+          // FIXME: enable unconditionally when software can set FIPS mode.
+          .fips_enable = !entropy_insert_enable,
           .route_to_firmware = false,
           .bypass_conditioner = false,
```
Comment (h-filali) [inline sw/device/tests/entropy_src_fw_observe_many_contiguous.c:209]: Can we also test the bypass mode? And what do you think about the route to SW?

**Review by h-filali — COMMENTED** [review]
Looks pretty darn good already. Thanks for working on this!
What do you think about adding some more HW entropy consumers?

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c [hunk 5]
```diff
@@ -158,15 +170,18 @@ static status_t entropy_config(
 status_t firmware_override_observe(
     uint32_t nr_samples, dif_entropy_src_single_bit_mode_t single_bit_mode,
     uint32_t timeout_usec, uint32_t repeat_count) {
+  LOG_INFO("==================");
```
[the remark this answers is not in the document]
Comment (jwnrt) [reply sw/device/tests/entropy_src_fw_observe_many_contiguous.c:243]: I think it was to more clearly separate each run for the `single_bit_mode`s (@pamaury)

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c [hunk 6]
```diff
@@ -103,9 +108,68 @@ static status_t let_observe_fifo_overflow(uint32_t timeout_usec) {
...
+
+  uint32_t buf[32];
+  size_t len = ARRAYSIZE(buf);
+  CHECK_DIF_OK(
+      dif_entropy_src_observe_fifo_nonblocking_read(&entropy_src, buf, &len));
+
+  if (len > 0) {
+    size_t words_written = 0;
+    // This is a non-blocking write that should be safe to run in this IRQ.
+    CHECK_DIF_OK(dif_entropy_src_observe_fifo_write(&entropy_src, buf, len,
```
[the remark this answers is not in the document]
Comment (jwnrt) [reply sw/device/tests/entropy_src_fw_observe_many_contiguous.c:133]: Oh good, I did think this was odd. I'm happy to rename this DIF in this PR, what would make a better name?

`dif_entropy_src_packer_fifo_write`?

I think @pamaury also wanted `_nonblocking` adding to the name, is that correct?

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (pamaury) [reply sw/device/tests/entropy_src_fw_observe_many_contiguous.c:243]: Yes, otherwise the various tests are hard to tell apart in the log. Feel free to change.

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (pamaury) [reply sw/device/tests/entropy_src_fw_observe_many_contiguous.c:133]: I think the name should mention that it is not blocking. Not sure about the name of the function, @h-filali do you have a suggestion?

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jwnrt) [reply sw/device/tests/entropy_src_fw_observe_many_contiguous.c:243]: I'm happy with it actually, I just moved it to above the `configure_entropy` call because I added some debug logs to that.

**Comment by jwnrt** [comment]
> What do you think about adding some more HW entropy consumers?

Yeah, that would be good. Do you have any suggestions? I'm actually aware of which blocks require entropy :sweat_smile:

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (h-filali) [reply sw/device/tests/entropy_src_fw_observe_many_contiguous.c:133]: Yes it would be great if you could do that @jwnrt .
@pamaury probably something like `dif_entropy_src_fw_ov_data_write` since the register is called [`fw_ov_wr_data`](https://opentitan.org/book/hw/ip/entropy_src/doc/registers.html#fw_ov_wr_data).

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c (hunk 4 above)
Comment (h-filali) [reply sw/device/tests/entropy_src_fw_observe_many_contiguous.c:209]: You know what, it's probably best to wait for @vogelpi to chip in here. I just read this:
https://github.com/lowRISC/opentitan/blob/050ca147cd71f13501cc4528fb37c2d4b9a8400a/hw/top_earlgrey/data/ip/chip_entropy_src_testplan.hjson#L242
and thought that this might be something we want.
In a nutshell you would set route_to_firmware to true and then the final entropy that comes out of the entropy_src wouldn't be routed to hardware but to firmware and it would then be readable via [ENTROPY_DATA](https://opentitan.org/book/hw/ip/entropy_src/doc/registers.html#entropy_data).
If you want to bypass the conditioner at the same time you can but you'd have to set [es_route](https://opentitan.org/book/hw/ip/entropy_src/doc/registers.html#entropy_control--es_type) to true as well.
You should be able to find a more detailed explanation [here](https://opentitan.org/book/hw/ip/entropy_src/doc/programmers_guide.html#firmware-override--bypass-modes). (I also made some additions to this section which are not merged yet [here](https://github.com/lowRISC/opentitan/pull/21528))

In conclusion I'd at least test it with the bypass mode for now and wait for what @vogelpi has to say in regards to the route_to_firmware.

**Comment by h-filali** [comment]
> > What do you think about adding some more HW entropy consumers?
> 
> Yeah, that would be good. Do you have any suggestions? I'm actually not very familiar with which blocks require entropy 😅

You may use any of the endpoints of the EDN I believe there are 8. Of the top of my head there is AES, KMAC, alert_handler, AST, OTBN urand, OTBN rand, IBEX. You may also draw inspiration from the powervirus test or from some of my edn tests.

File: sw/device/tests/entropy_src_fw_observe_many_contiguous.c (hunk 4 above)
Comment (vogelpi) [reply sw/device/tests/entropy_src_fw_observe_many_contiguous.c:209]: We should test the bypass mode (conditioner off) but I'll comment in a different place regarding this.

We don't need to test the `.route_to_firmware` option here. This is a separate feature in the testplan and the main purpose of this feature (AFAIK) is to test the conditioner in isolation: i.e., you write a known value into the preconditioner FIFO, then let the conditioner process and read the produced output seed from ENTROPY_DATA. This is also known as known answer test (KAT) and it's what the existing test `chip_sw_entropy_src_known_answer_tests` does. Disabling the conditioner in this test would be possible but it's kind of the next step when debugging the test in order to find out if at least the path FW_OV_WR_DATA -> Bypass FIFO -> ENTROPY_DATA works. I think this is so basic that it doesn't require a test.

**Review by vogelpi — COMMENTED** [review]
Thanks @jwnrt for working on this!

I think this is going in the right direction but a couple of things should be changed before we merge this test. I am giving some more detailed feedback and explanations inline below.

> 1. Adds an interrupt handler for the `es_observe_fifo_ready` interrupt which reads 32 words (the FIFO threshold value) from the observe FIFO and writes it back again.
> 
I think this is perfectly fine for a proof of concept but we will want to read at least 1024 consecutive samples, i.e., the window (can be 4-bit samples, i.e., 4 kByte of data or 1-bit samples in single-bit mode). The reason is that firmware will need to read a consecutive chunk of entropy for the health testing (done by firmware in this mode but obviously not part of this test). Similar to the first test @pamaury wrote, you'll need to:
- test that the FIFO doesn't overflow while reading the 1024 samples
- when it overflows, the FIFO is emptied before starting collecting entropy for the next window
- do a couple of iterations/windows without restarting the entropy source in between, to stress test the whole entropy complex. For this, you can configure the EDNs in auto mode and request a lot of entropy (@h-filali knows how to achieve this).

> ## TODO and questions
> ### FIPS mode
> 
> Currently, the test has to disable FIPS mode when running with extract and insert. With FIPS enabled, CSRNG never instantiates. This is likely due to software not being able to set the FIPS bit which CSRNG probably expects, fixed by #21369.

I think this is unrelated. The problem with FIPS mode is that the data you write into the FW_OV_WR_DATA register will be injected into the conditioner. Firmware also has to kick off the conditioner. If this is not done, ENTROPY_SRC never generates a seed and that's why CSRNG fails to instantiate: it doesn't get the seed for the instantiation. How to operate the conditioner in firmware override: extract and insert mode is described here: https://opentitan.org/book/hw/ip/entropy_src/doc/programmers_guide.html#firmware-override---extract--insert

The second challenge here is that whether a seed is / can be produced depends on the amount of data you insert: without conditioning it's 384 bits or 12 32-bti words, with conditioning it may be more. When fully relying on the hardware pipeline, we use 2048 bits per default for one 384-bit seed. So that would be the content of the full observe FIFO. I think right now, you're at most writing 1024 bits.

In general the amount of data the test extracts and then inserts doesn't need to match. It's fine to extract different amounts than what you insert. Firmware would most likely do this as well.

> ### Checking the output of the entropy source
> 
> Is generating random numbers the best way to check the output of the entropy source? Is there something else that would be better to run? I've checked that when the interrupt handler is not writing data back into the FIFO the random number generators time out. Should I try incorporate that into the test somehow?

Yes, it's intended to check various hardware entropy consumers are run and indeed finish their operation as @h-filali mentioned. One source of inspiration is the power virus test. IIRC, Hakim also did something similar but simpler for one of the EDN SiVal tests.

Ideally, you can also extend this test to check the [DEBUG_STATUS](https://opentitan.org/book/hw/ip/entropy_src/doc/registers.html#debug_status) before injecting a seed. This has a field ENTROPY_FIFO_DEPTH indicating if the final FIFO in the pipeline is empty (CSRNG stalls) or full (any seed injected by firmware gets dropped).

Sorry for the long message. I'll be OoO tomorrow and wanted to be sure you're not blocked on my input :-)

**Comment by pamaury** [comment]
For what it is worth, I don't think it is necessary to read 1024 samples in the ISR. It seems perfectly fine to me to read, for example, 512 samples the first time it fires and then 512 the second time. The ISR can check if the FIFO overflow in the mean time so it can still determine if it got contiguous samples or not.
In fact, waiting until we have 1024 samples seem very inefficient because now we won't start the conditioner until after collection instead of slowly pushing the data through the conditioner as it becomes available.

**Comment by jwnrt** [comment]
I guess the problem is that if we push 512 bytes on the first interrupt and then get an overflow, we can't take back those 512 bytes that we already sent through (I don't think). That would mean that the next 512 bytes get added on top and there's no way to tell the downstream consumers that what they're seeing is two sets of 512 bytes and not one contiguous set of 1024. Is that correct?

**Comment by pamaury** [comment]
This is why the reading and writing paths need to be asynchronous: there is one big 1024-sample buffer. The reading ISR regularly reads samples from the observe FIFO and writes them to the buffer. It needs to do a small amount of work each time to not block the rest of the system (so read way less than 512 samples). Concurrently, probably using a timer, another ISR pushes data from the buffer into the conditioner when there is space available; most importantly it does not block.
This is arguably a more complex setup but it is also more efficient and more representative of an actual workload in my opinion.

**Comment by jwnrt** [comment]
Just so I'm understanding, are you saying to have the read ISR push small amounts of data into an in-memory buffer so that the observe FIFO doesn't overflow, then have another ISR on a timer which takes chunks of 1024 bytes from that buffer and writes them through to the FIFO that does conditioning and the rest of the entropy source chain?

**Comment by jwnrt** [comment]
I tried that with an ISR that blocked and wrote the whole buffer of 1024 to the conditioner in one go and it was too slow. I could try again with an ISR that writes small chunks at a time, but only starts once 1024 contiguous samples are there. The implementation might be a bit more complex, but I'll see if it's doable.

**Comment by jwnrt** [comment]
I've updated the implementation to read and write 32 words at a time in each interrupt while still handling overflows. This runs a lot faster. It's triggered by the "entropy FIFO full" and "entropy requested" interrupts, but I can try and change it to a timer if needed (setting up a repeating timer ended up being more complicated than I expected).

I've also separated this test off from the observe test because it was starting to get quite complex.

I still need to add a bunch more entropy consumers.

**Comment by jwnrt** [comment]
Sorry for the delay, I've added some usage of AES, KMAC, and HMAC. Please let me know if my usage of them isn't sufficient. I didn't copy all of the things that the power virus test does because it's quite complex and a lot would need to be moved across. The results of these processes aren't checked, just the statuses.

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+bool entropy_src_fifo_has_overflowed(void) {
+  uint32_t fifo_depth;
+  CHECK_DIF_OK(dif_entropy_src_get_fifo_depth(&entropy_src, &fifo_depth));
+  return fifo_depth == ENTROPY_SRC_PARAM_OBSERVE_FIFO_DEPTH;
+}
+
+/**
+ * Drain observe FIFO.
+ */
+static void drain_observe_fifo(void) {
```
Comment (pamaury) [inline sw/device/tests/entropy_src_fw_override_test.c:111]: maybe not necessary for this PR but we could consider moving to the testutils now that it is used in two tests

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+  // This value is arbitrary, it could be 1 but since there is some
+  // overhead in dif_entropy_src_observe_fifo_nonblocking_read, it's better
+  // to read several words every time to drain the FIFO quickly.
+  const size_t kDrainCount = 32;
+  size_t len;
+  // Read from the FIFO until we get a short read which means that the FIFO was
+  // emptied.
+  LOG_INFO("drain observe FIFO overflow...");
```
Comment (pamaury) [inline sw/device/tests/entropy_src_fw_override_test.c:119]: probably should not log in this function because it can be called in ISR context

Comment (pamaury) [reply sw/device/tests/entropy_src_fw_override_test.c:119]: If you really do want to report that an overflow occured then maybe have a global variable counting the stats (e.g. number of overflows) and display that at the end of the test?

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+    drain_observe_fifo();
+    // Reset the input buffer.
+    input_buf -= (kFifoBufferSizeWords - words_to_input);
+    words_to_input = kFifoBufferSizeWords;
+  }
+
+  if (words_to_input > 0) {
+    // Move at most 32 samples into the input buffer.
+    size_t len = words_to_input;
+    if (len > 32) {
```
Comment (pamaury) [inline sw/device/tests/entropy_src_fw_override_test.c:149]: Maybe create a constant for this?

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+    words_to_input = kFifoBufferSizeWords;
+    words_to_output = kFifoBufferSizeWords;
+  }
+
+  if (words_to_output > 0) {
+    // Move out at most 16 samples into the override FIFO.
+    // This number has to be small otherwise the conditioner runs for too long
+    // and we spend too much time in this ISR.
+    size_t len = words_to_output;
+    if (len > 16) {
```
Comment (pamaury) [inline sw/device/tests/entropy_src_fw_override_test.c:177]: create a constant for this? (and make sure to add the comment as well)

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+      len = 16;
+    }
+
+    // Start the SHA3 block ready for us to start inserting into the FIFO.
```
Comment (pamaury) [inline sw/device/tests/entropy_src_fw_override_test.c:181]: Unclear sentence

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+        &entropy_src, kDifToggleEnabled));
+
+    size_t words_written = 0;
+    CHECK_DIF_OK(dif_entropy_src_fw_ov_data_write(&entropy_src, output_buf, len,
+                                                  &words_written));
+
+    output_buf += words_written;
+    words_to_output -= words_written;
+
+    // Disable the SHA3 block, allowing it to finish processing and pass the
```
Comment (pamaury) [inline sw/device/tests/entropy_src_fw_override_test.c:192]: nit: I don't think this technically disables the SHA3 block, rather as you wrote it triggers the SHA3 conditioner (and disable the insertion mecanism).

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+    bool bypass_conditioner) {
+  // Don't let the extract and insert interrupt handler run until we've enabled
+  // the entropy source.
+  irq_external_ctrl(false);
+
+  LOG_INFO("Disabling entropy complex");
+  // Disable the entropy complex.
+  TRY(entropy_testutils_stop_all());
+
+  // Disable all health tests.
```
Comment (pamaury) [inline sw/device/tests/entropy_src_fw_override_test.c:234]: Maybe we should move this sequence to the test utils as well? It's useful and not complete obvious

**Review by pamaury — APPROVED** [review]
Really nice @jwnrt :) I left a couple of minor comments.

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+  LOG_INFO("Configuring interrupts...");
+  configure_interrupts();
+  irq_global_ctrl(true);
+  irq_external_ctrl(false);
+
+  LOG_INFO("Computing EDN parameters");
+  edn_params0 = edn_testutils_auto_params_build(
+      /*disable_rand=*/false, /*res_itval=*/0, /*glen_val=*/0);
+  edn_params1 = edn_testutils_auto_params_build(
+      /*disable_rand=*/false, /*res_itval=*/0, /*glen_val=*/0);
```
Comment (vogelpi) [inline sw/device/tests/entropy_src_fw_override_test.c:388]: I think for this test we should use:
```
/*disable_rand=*/true, /*res_itval=*/1, /*glen_val=*/1)
```
This will configure the EDNs/CSRNG instances in a mode where they request maximum entropy from ENTROPY_SRC :-)

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+                                            &data));
+  }
+
+  // Run an AES encryption and decryption process.
+  dif_aes_transaction_t transaction = {
+      .operation = kDifAesOperationEncrypt,
+      .mode = kDifAesModeEcb,
+      .key_len = kDifAesKey256,
+      .key_provider = kDifAesKeySoftwareProvided,
+      .mask_reseeding = kDifAesReseedPerBlock,
```
Comment (vogelpi) [inline sw/device/tests/entropy_src_fw_override_test.c:314]: This is already nice!

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 16]
```diff
@@ -0,0 +1,407 @@
...
+      .manual_operation = kDifAesManualOperationAuto,
+      .reseed_on_key_change = false,
```
Comment (vogelpi) [inline sw/device/tests/entropy_src_fw_override_test.c:316]: Change this to true to use more entropy :-)

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 17]
```diff
@@ -0,0 +1,407 @@
...
+  // Run KMAC
+  dif_kmac_key_t software_key = {
+      .share0 = {0x43424140, 0x47464544, 0x4B4A4948, 0x4F4E4D4C, 0x53525150,
+                 0x57565554, 0x5B5A5958, 0x5F5E5D5C},
+      .share1 = {0},
+      .length = kDifKmacKeyLen256,
+  };
+  uint32_t output[8];
+
+  TRY(kmac_testutils_config(&kmac, false));
```
Comment (vogelpi) [inline sw/device/tests/entropy_src_fw_override_test.c:334]: The default KMAC config will not use any entropy from EDN. This is unfortunate. Is it possible to make the following changes here?
```
.entropy_mode = kDifKmacEntropyModeEdn,
.entropy_fast_process = false,
.entropy_hash_threshold = 1,
.msg_mask = true,
```

File: sw/device/tests/entropy_src_fw_override_test.c [hunk 18]
```diff
@@ -0,0 +1,407 @@
...
+      .message_endianness = kDifHmacEndiannessLittle,
+  };
+
+  TRY(dif_hmac_mode_sha256_start(&hmac, hmac_transaction));
+  TRY(hmac_testutils_push_message(&hmac, (char *)kHmacRefLongKey,
+                                  sizeof(kHmacRefLongKey)));
+  TRY(hmac_testutils_check_message_length(&hmac, sizeof(kHmacRefLongKey) * 8));
+  TRY(dif_hmac_process(&hmac));
+  dif_hmac_digest_t digest;
+  TRY(hmac_testutils_finish_polled(&hmac, &digest));
```
Comment (vogelpi) [inline sw/device/tests/entropy_src_fw_override_test.c:352]: Unfortunately HMAC doesn't have an EDN connection. So won't request entropy. Instead, it would e good to run a simple test on OTBN which has 2 EDN interfaces :-)

From the `edn_auto_mode.c` test, I think you should be able to use this function `otbn_randomness_test_prepare(&otbn, kOtbnRandomnessIterations);` to load a program that will use quite some entropy on both interfaces. Then you can run it and wait for completion.

**Review by vogelpi — APPROVED** [review]
Very nice work @jwnrt ! Thanks for doing this.

Besides mostly minor comments, I have two things:
1. Ideally some more entropy can be injected into the conditioner but I see the issue you raised. So lets maybe not do it in this iteration.
2. HMAC doesn't use entropy from EDN but OTBN would use both EDNs. I suggest to execute the `otbn_randomness_test`.

File: sw/device/tests/entropy_src_fw_override_test.c (hunk 8 above)
Comment (jwnrt) [reply sw/device/tests/entropy_src_fw_override_test.c]: I've moved the log outside of the function, but haven't added a global counter even though that could be useful for debugging this test.

File: sw/device/tests/entropy_src_fw_override_test.c (hunk 11 above)
Comment (jwnrt) [reply sw/device/tests/entropy_src_fw_override_test.c]: I've tried to make these two comments clearer, let me know what you think

**Comment by vogelpi** [comment]
Awesome, thanks @jwnrt !

**Comment by vogelpi** [comment]
This resolves [#20952](https://github.com/lowRISC/opentitan/issues/20952).

**Comment by jwnrt** [comment]
I think the changes that will be needed for the backport are:

1. The workaround for the observe FIFO overflow will need to be added back.
2. OTBN required the FIPS bit to be set by software which isn't possible on earlgrey_es_sival.

If the OTBN issue cannot be worked around then we may have to depend on the other entropy consumers.

**Comment by vogelpi** [comment]
Your assessment is correct @jwnrt . We won't be able to use OTBN on earlgrey_es_sival for this test. Maybe we can do a couple of iteration of AES and KMAC instead.
